### PR TITLE
WC-23 flex

### PR DESCRIPTION
**Pull request by eilinora**
#### Related issues

Fixes https://meetup.atlassian.net/browse/WC-23

#### Description
DoD:
- Flex & FlexItem components have stories, & tests

#### Screenshots (if applicable)
screen shot 2017-01-23 at 12 24 25 pm
screen shot 2017-01-23 at 12 24 22 pm

### CODE REVIEW

File: src/Flex.jsx [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+				[FLEX_NOGUTTER_CLASS]: noGutters,
+				[`${FLEX_CLASS}--${VALID_SPACE[justify]}`]: justify,
+				[`${FLEX_ALIGN_CLASS}${VALID_ALIGNMENTS[align]}`]: align,
+			}, className);
+
+		return (
+			<div
+				className={classNames}
+				{...other}>
+					{children}
```
Comment (mmcgahan) [inline src/Flex.jsx]: this is indented on more than it needs to be - prefer putting the `<div`-closing `>` on a newline to make the attribute/prop nesting more clear

File: src/flex.story.jsx [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+	.add('No gutters', () => (
+		<Flex noGutters style={flexParentStyles}>
+			<FlexItem style={flexItemStyles}><div style={boxStyles}>Item 1</div></FlexItem>
+			<FlexItem style={flexItemStyles}><div style={boxStyles}>Item 2</div></FlexItem>
+			<FlexItem style={flexItemStyles}><div style={boxStyles}>Item 3</div></FlexItem>
+			<FlexItem style={flexItemStyles}><div style={boxStyles}>Item 4</div></FlexItem>
+			<FlexItem style={flexItemStyles}><div style={boxStyles}>Item 5</div></FlexItem>
+		</Flex>
+	))
+	.add('Justify center', () => (
```
Comment (akdetrick) [inline src/flex.story.jsx]: Not a blocker, but the justification change isn't really visible in storybook right now; We may need to add a parent container around `Flex` here to visually demonstrate `justify-content` variants

**Comment by akdetrick** [comment]
Aside from my comment about showing `justify-content` variants in storybook, this LGTM. 

Merge at will.

File: src/flex.test.jsx [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+describe('Flex', function() {
+
+	it('exists', function() {
+		const flex = TestUtils.renderIntoDocument(<Flex />);
+		const flexNode = ReactDOM.findDOMNode(flex);
+
+		expect(flexNode).not.toBeNull();
+	});
+
```
Comment (nlaz) [inline src/flex.test.jsx]: nitpick request: keep the spacing in this file consistent. Some `describe`s and `it`s have newlines after them and some don't. Keeping it consistent will make it look cleaner and will soothe my OCD.

File: src/Flex.jsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
+import React from 'react';
+import cx from 'classnames';
+
+
```
Comment (nlaz) [inline src/Flex.jsx]: nitpick: remove extra newline

File: src/Flex.jsx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+		const columnReverseBreakpoint = VALID_BREAKPOINTS[columnReverse] || VALID_BREAKPOINTS['all'];
+		const rowReverseBreakpoint = VALID_BREAKPOINTS[rowReverse] || VALID_BREAKPOINTS['all'];
+
+		const isColumn = direction === DIRECTION_COLUMN;
+		const classNames = cx(
+			FLEX_CLASS,
+			{
+				// horizontal default
+				[FLEX_ROW_CLASS]: !isColumn,
```
Comment (nlaz) [inline src/Flex.jsx]: I know the square brackets are not needed for plain strings like `FLEX_ROW_CLASS`. Is this just a style thing to keep it consistent with the interpolated strings? I can appreciate that.

File: src/Flex.jsx [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+				// reverse breakpoint modifiers
+				[`${rowReverseBreakpoint}_flex--rowReverse`]: rowReverse,
+				[`${columnReverseBreakpoint}_flex--columnReverse`]: columnReverse,
+
+				// other
+				[FLEX_WRAP_CLASS]: wrap,
+				[FLEX_NOGUTTER_CLASS]: noGutters,
+				[`${FLEX_CLASS}--${VALID_SPACE[justify]}`]: justify,
+				[`${FLEX_ALIGN_CLASS}${VALID_ALIGNMENTS[align]}`]: align,
```
Comment (nlaz) [inline src/Flex.jsx]: There are a lot of classes and modifiers. You did a good job here handling them 👍

File: src/Flex.jsx [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+	rowReverse: React.PropTypes.oneOfType([
+		React.PropTypes.bool,
+		React.PropTypes.oneOf(Object.keys(VALID_BREAKPOINTS))
+	]),
+	columnReverse: React.PropTypes.oneOfType([
+		React.PropTypes.bool,
+		React.PropTypes.oneOf(Object.keys(VALID_BREAKPOINTS))
+	]),
+
```
Comment (nlaz) [inline src/Flex.jsx]: another spacing nitpick: kill the extra newlines in this object?

File: src/FlexItem.jsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+			...other
+		} = this.props;
+
+		const classNames = cx(
+			FLEX_ITEM_CLASS,
+			{
+				[FLEX_ITEM_SHRINK_CLASS]: shrink,
+				[`${FLEX_ITEM_GROW_CLASS}${growFactor}`] : growFactor,
+			},
+			className
```
Comment (nlaz) [inline src/FlexItem.jsx]: Another nitpick: I am not a fan of the large object block in the middle of the classNames function. I prefer something more compact like:

```js
const classNames = cx(
   FLEX_ITEM_CLASS,
   { FLEX_ITEM_SHRINK_CLASS: shrink },
   { [`${FLEX_ITEM_GROW_CLASS}${growFactor}`]: growFactor },
   className
);
```
This is a personal preference. How you have it is completely fine as well.

File: src/flex.story.jsx [hunk 9]
```diff
@@ -0,0 +1,215 @@
+
```
Comment (nlaz) [inline src/flex.story.jsx]: nitpick: kill this newline

File: src/Flex.jsx (hunk 5 above)
Comment (eilinora) [inline src/Flex.jsx]: FLEX_ROW_CLASS is a constant i'm using ast the prop name

**Comment by eilinora** [comment]
Found a bug with justify, docs wasn't the same as css names
screen shot 2017-01-24 at 12 27 55 pm

File: src/Flex.jsx (hunk 5 above)
Comment (nlaz) [inline src/Flex.jsx]: ooh interesting. that makes sense because JS objects can have fields without quotation marks. nevermind then!

**Pull request merged**